packages/@aws-cdk/cdk/lib/cross-environment-token.ts

```diff
@@ -0,0 +1,66 @@
+import { ArnComponents } from './arn';
+import { DeployTimeOrAssignedPhysicalName } from "./deploy-time-or-assigned-physical-name";
+import { PhysicalName } from './physical-name';
+import { IResource } from './resource';
+import { ResolveContext, Token } from './token';
+
+/**
+ * A Token that represents a reference that spans accounts and/or regions,
+ * and so requires the resources to have physical names.
+ * You should never need to interact with these directly,
+ * instead use the {@link ResourceIdentifiers} class.
+ * This class is private to the @aws-cdk/cdk package.
+ */
+export abstract class CrossEnvironmentToken extends Token {
+  private readonly resource: IResource;
+  private readonly physicalName: PhysicalName;
+
+  /**
+   * @param regularValue the value used when this is referenced NOT from a cross account and/or region Stack
+   * @param crossEnvironmentValue the value used when this is referenced from a cross account and/or region Stack
+   * @param resource the scope this reference is mastered in. Used to determine the owning Stack
+   * @param physicalName
+   * @param displayName a short name to be used in Token display
+   */
+  protected constructor(private readonly regularValue: string, private readonly crossEnvironmentValue: any,
+                        resource: IResource, physicalName: PhysicalName, displayName: string) {
+    super(undefined, displayName);
+
+    this.resource = resource;
+    this.physicalName = physicalName;
+  }
+
+  public resolve(context: ResolveContext): any {
+    const consumingStack = context.scope.node.stack;
+    const owningStack = this.resource.node.stack;
+
+    if (consumingStack.env.account !== owningStack.env.account ||
+        consumingStack.env.region !== owningStack.env.region) {
+      // check whether the resource has a physical name set
+      if (!this.physicalName.availableAtSynthesisTime) {
+        if (this.physicalName instanceof DeployTimeOrAssignedPhysicalName) {
```

Comment:
Try to avoid instanceof (theoretically it might not work in multi-versioned situation, which I know we don't really have, but we try to avoid it anyway)

Reply:
I've removed this code with the other refactoring, so this is no longer an issue.